lib/ansible/module_utils/urls.py

```diff
@@ -1574,8 +1574,15 @@ def fetch_file(module, url, data=None, headers=None, method=None,
     '''
     # download file
     bufsize = 65536
-    file_name, file_ext = os.path.splitext(str(url.rsplit('/', 1)[1]))
-    fetch_temp_file = tempfile.NamedTemporaryFile(dir=module.tmpdir, prefix=file_name, suffix=file_ext, delete=False)
+    # We try to preserve all filename extensions here to preserve things like
+    # .tar.gz so that code that makes assumptions about a file's contents based
+    # on whether a filename has multiple extensions doesn't get confused.
+    file_name = str(url.rsplit('/', 1)[1])
+    file_parts = file_name[1:].split('.', 1)  # [1:] avoids splitting ".hiddenfile"
+    file_exts = "." + file_parts[1] if len(file_parts) > 1 else ""
+    if file_exts:
+        file_name = file_name[:len(file_exts) - 1]
```

Comment:
```suggestion
    file_name, file_exts = os.path.splitext(str(url.rsplit('/', 1)[1]))
    # Preserving double filename extensions like .tar.gz
    _, double_ext = os.path.splitext(file_name)
    if double_ext:
        file_exts = double_ext + file_exts:
```

Reply:
Your suggested change has a couple of issues that I see. First, although it makes `fetch_file` avoid breaking filename extensions for double extensions, my change is more robust since it supports an arbitrary number of filename extensions. Second, my change only changes the point in the filename where the randomized string is inserted (before all extensions), whereas your change will change the that point like mine, but also leave `double_ext` appearing twice in the filename (not going break anything I would assume, but it seems a little weird/sloppy).